File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        # Brute force solution - O(N^3) - Time Limit Exceeded
        # ans = list()
        # n = len(nums)
        # for i in range(n):
        #     for j in range(i+1,n):
        #         for k in range(j+1, n):
        #             if nums[i] + nums[j] + nums[k] == 0 and sorted([nums[i], nums[j], nums[k]]) not in ans:
        #                 ans.append(sorted([nums[i], nums[j], nums[k]]))

        # return ans

        # 3 pointer approach
        n = len(nums)
        nums.sort()
        ans = list()
        for i in range(n):

            # remove duplicates
            if i != 0 and nums[i] == nums[i-1]:
                    continue 

            left = i + 1
            right = n - 1

            while left < right:

                total_sum = nums[i] + nums[left] + nums[right]
                
                if total_sum < 0:
                    left += 1
                
                elif total_sum > 0:
                    right -= 1

                else:
                    ans.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1

                    # skipping the duplicate values
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    
                    while left < right and  nums[right] == nums[right + 1]:
                        right -= 1

        return ans

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    
    let n = nums.length
    let answer = []
    
    let compareFn = (a,b) => {
        if (a<b){
            return -1
        }
        else if (a>b){
            return 1
        }
        else{
            return 0
        }
    }
    
    nums.sort(compareFn)
    
    let i = 0
    while (i < n-1){
        let j = i + 1
        let k = n - 1
        
        if (nums[i] > 0){
            i += 1
            continue
        }
        
        while (j < k){
            let sum_ = nums[i] + nums[j] + nums[k]
            
            if (sum_ == 0){
                answer.push([nums[i], nums[j], nums[k]])
                
                while (nums[j] === nums[j+1]){
                    j += 1
                }
                
                while (nums[k] === nums[k-1]){
                    k -= 1
                }
                
                j += 1
                k -= 1
                 
            }
            else if (sum_ > 0){
                k -= 1
            }
            else{
                j += 1
            }
        }
        
        while (i < n - 1 && nums[i] === nums[i+1]){
            i += 1
        }
        
        i += 1
    }
    
    return answer
};